api/envoy/config/filter/http/rbac/v2/rbac.proto

```diff
@@ -25,6 +26,9 @@ message RBAC {
   // but will emit stats and logs and can be used for rule testing.
   // If absent, no shadow RBAC policy will be applied.
   config.rbac.v2.RBAC shadow_rules = 2;
+
+  // A symbolic expression over request attributes specifying an access control condition.
+  google.api.expr.v1alpha1.Expr condition = 3;
```

Comment:
Can we add this to the `Policy` in the `config.rbac.v2.RBAC`? So that you can just extend the RBAC engine to evaluate the `condition` in addition to the `permission` and `principal`. This will also support to use it for TCP protocol and in per-route rules.

Reply:
+1, it will also be separate expression for rules and shadow_rules.

Reply:
The CEL expression is a heavy object. It's preferable to have a single condition with ANDs rather than many small expressions.

Reply:
Can we move it to RBAC proto as a middle-ground? It will respect the action and shadowing, but not have the outer policy_id structure.

Reply:
> Can we move it to RBAC proto as a middle-ground? It will respect the action and shadowing, but not have the outer policy_id structure.

I'm OK with this, having shadow_rules support is great.

Reply:
Putting inside the `Policy` gives us more flexibility. Yes the CEL is a heavy object but it's always up to the control plane to decide how to use it. Even putting inside the Policy, the control plane could still choose to combine all expressions into a single CEL object. On the other hand, if the control plane want to have several CEL objects for some reasons, they still have the ability to do this.

The `effective_policy_id` is generated from the map key which is quite useful in monitoring and also shadow rule, it tells you which actual policy rejects the request.

And last I think this is also going to simplify the code in `engine_impl.cc` a lot as you can just modify the https://github.com/envoyproxy/envoy/blob/b22d2b5cf09f779962cfedaaab24969f384cbc48/source/extensions/filters/common/rbac/matchers.cc#L162-L163 to add a statement like `&& condition_.matches()`.

Reply:
Updated per the comment. Signatures needed to be updated since stream info is needed.